qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py

```diff
@@ -0,0 +1,256 @@
+
+    with_labels: bool = False
+
+    node_list: Optional[list] = None
+
+    edge_list: Optional[list] = None
+
+    node_size: Union[int, list] = 300
+
+    node_color: Union[
+        str,
+        Tuple[float, float, float],
+        Tuple[float, float, float],
+        List[Union[str, Tuple[float, float, float], Tuple[float, float, float, float]]],
+    ] = "#1f78b4"
+
+    node_shape: str = "o"
+
+    # node and edge transparency
+    alpha: Optional[float] = None
+
+    # Matplotlib colormap object
+    cmap: Optional["Colormap"] = None
+
+    # minimum value for node colormap scaling
+    vmin: Optional[float] = None
+
+    # minimum value for node colormap scaling
+    vmax: Optional[float] = None
+
+    linewidths: Union[float, Sequence] = 1.0
+
+    width: Union[float, Sequence] = 1.0
+
+    edge_color: Union[str, Sequence] = "k"
+
+    edge_cmap: Optional["Colormap"] = None
+
+    # minimum value for edge colormap scaling
+    edge_vmin: Optional[float] = None
+
+    # maximum value for node colormap scaling
+    edge_vmax: Optional[float] = None
+
+    style: str = "solid"
+
+    labels: Optional[Callable] = None
+
+    edge_labels: Optional[Callable] = None
+
+    font_size: int = 12
+
+    font_color: str = "k"
+
+    font_weight: str = "normal"
+
+    font_family: str = "sans-serif"
+
+    label: Optional[str] = None
+
+    connectionstyle: str = "arc3"
+
+
+class Lattice:
+    """General Lattice."""
+
+    def __init__(self, graph: PyGraph) -> None:
+        """
+        Args:
+            graph: Input graph for Lattice. `graph.multigraph` must be False.
+
+        Raises:
+            ValueError: If `graph.multigraph` is True for a given graph, it is invalid.
+        """
+        if graph.multigraph:
+            raise ValueError(
+                f"Invalid `graph.multigraph` {graph.multigraph} is given. "
+                "`graph.multigraph` must be `False`."
+            )
+        if graph.edges() == [None] * graph.num_edges():
+            weighted_edges = [edge + (1.0,) for edge in graph.edge_list()]
+            for start, end, weight in weighted_edges:
+                graph.update_edge(start, end, weight)
+        self._graph = graph
+
+        self.pos: Optional[dict] = None
+
+    @property
+    def graph(self) -> PyGraph:
+        """Return a copy of the input graph."""
+        return self._graph.copy()
+
+    @property
+    def num_nodes(self) -> int:
+        """Return the number of nodes."""
+        return self.graph.num_nodes()
+
+    @property
+    def node_indexes(self) -> NodeIndices:
+        """Return the node indexes."""
+        return self.graph.node_indexes()
+
+    @property
+    def weighted_edge_list(self) -> WeightedEdgeList:
+        """Return a list of weighted edges."""
+        return self.graph.weighted_edge_list()
+
+    def copy(self) -> "Lattice":
+        """Return a copy of the lattice."""
+        return Lattice(self.graph.copy())
```

Comment:
Since self.graph already does a copy() this seems to be doing a copy of the copy. What is this expected to be used for - I am asking since it seems to be in this base class but not overridden (that I saw) in any derived classes. Hence if I call copy() on a derived instance, say a HyperCubicLattice, I will not get a copy of that rather I am going to get a base class Lattice instance using its graph. Since  user can make a Lattice from the graph property easily themselves too, with the above observations I guess I am questioning whether having this method is useful

Reply:
In thinking/reviewing a little more I wondered about the sub-classes since it seemed they just created a graph for the Lattice constructor and only a couple had a little more. I was wondering if all those classes, all those types, were actually needed/useful - rather than say just have factory functions that build out a Lattice in square, triangular form etc. In a couple of the cases it seems that the parameters passed to the constructor are used in the draw without boundary. Otherwise these sub-classes do not add any new/specialized functions at for the derived types. Like for instance being able to add one more row to the lattice. Now I see there are public variables for rows, cols etc in sub-classes. While I can see reading these as properties can be useful when there are these subtypes, they are public instance vars, not properties like is done in Lattice, and thus they can be set too. Which will either have no effect (square lattice) or later mess up the draw without boundary. Should all these be private instance vars and have properties just to read them from a public API perspective. Maybe these sub-types can be expanded upon in the future so it seems more useful to have them as specific types - at present, to me, aside from draw without boundary, that maybe cannot be done in Lattice itself from the graph, their existence just adds a bunch of types/classes with no other additional specilalization/value beyond that.

Reply:
I think this is correct about copy. (or copy.deepcopy might be safer.)
https://github.com/Qiskit/qiskit-nature/pull/365/commits/8beeb63c304d5bbcc1009727fd1b4f0fa68c0c79#diff-f5c282fdc2a29ac3872af2c25326ad98d8277544a967e89fb6dec888b872c5c9R151

For public API, I'm sure there were unnecessary setters, so I deleted it and made them getters only.
8af7388

Currently, there is no special method, but the meaning is clear from the class name, so I believe it is easier to use than, for example, creating from a class method.

Reply:
> For public API, I'm sure there were unnecessary setters, so I deleted it and made them getters only.

For the hyper_cubic_lattice is still seems to have a `dim` and `size` public instance vars.

> Currently, there is no special method, but the meaning is clear from the class name, so I believe it is easier to use than, for example, creating from a class method.

Ok, I was more thinking of public methods say `square_lattice(.....) -> Lattice`. But perhaps, even though the various types today do not have any specialization beyond the one method in a couple of the classes, in the future that will happen and then the structure is more suitable for this than say just having methods.

Reply:
@woodsp-ibm thanks for the comment. In general I agree with @ikkoham that the current implementation is more user friendly and the class names correspond to the expected lattice.

Reply:
I forgot `dim` and `size`. I'll add them soon.

As for type hint, they are fine for returning sub classes. (Liskov) If you need in the future, we can write stronger type hints in the sub class side.

Reply:
For copy() that's fine - what you have now is a bit different than before in that now its making a copy of the instance, and hence whatever the subclass is, not a new Lattice (possibly parent of instance) with the same graph.